src/auxinfo/proof.rs

```diff
@@ -142,3 +183,123 @@ impl AuxInfoProof {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{paillier::prime_gen, protocol::SharedContext, utils::testing::init_testing};
+    use rand::Rng;
+
+    fn random_auxinfo_proof<R: RngCore + CryptoRng>(
+        _rng: &mut R,
+        test_code: impl FnOnce(CommonInput, AuxInfoProof) -> Result<()>,
+    ) -> Result<()> {
+        let mut rng = init_testing();
+        let sid = Identifier::random(&mut rng);
+        let rho = rng.gen();
+        let setup_params = VerifiedRingPedersen::gen(&mut rng, &()).unwrap();
+        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng).unwrap();
+        let modulus = &p * &q;
+        let shared_context = SharedContext::random(&mut rng);
+        let common_input = CommonInput::new(&shared_context, sid, rho, &setup_params, &modulus);
+        let proof = AuxInfoProof::prove(
+            &mut rng.clone(),
+            &common_input.clone(),
+            &p.clone(),
+            &q.clone(),
+        )
+        .unwrap();
+        //(Ok(proof.clone()), Ok(common_input.clone()))
+        test_code(common_input, proof)
+    }
+
+    #[test]
+    fn auxinfo_proof_verifies() -> Result<()> {
+        let mut rng = init_testing();
+        let sid = Identifier::random(&mut rng);
+        let rho = rng.gen();
+        let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
+        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng).unwrap();
+        let modulus = &p * &q;
+        let shared_context = SharedContext::random(&mut rng);
+        let common_input = CommonInput::new(&shared_context, sid, rho, &setup_params, &modulus);
+        let proof = AuxInfoProof::prove(&mut rng, &common_input, &p, &q)?;
+        let common_input = CommonInput::new(&shared_context, sid, rho, &setup_params, &modulus);
+        assert!(proof.verify(&common_input).is_ok());
+        Ok(())
+    }
+
+    #[test]
+    fn each_constituent_proof_must_be_valid() -> Result<()> {
+        let mut rng = init_testing();
+        let mut rng_clone = rng.clone();
+        let f = |input: CommonInput, proof: AuxInfoProof| {
+            let f1 = |input1: CommonInput, proof1: AuxInfoProof| {
+                let mix_one = AuxInfoProof {
+                    pifac: proof.pifac,
+                    pimod: proof1.pimod,
+                };
+                let mix_two = AuxInfoProof {
+                    pifac: proof1.pifac,
+                    pimod: proof.pimod,
+                };
+                assert!(mix_one.verify(&input).is_err());
+                assert!(mix_two.verify(&input1).is_err());
+                Ok(())
+            };
+            let _ = random_auxinfo_proof(&mut rng_clone, f1);
+            Ok(())
+        };
+        let _ = random_auxinfo_proof(&mut rng, f);
+        Ok(())
```

Comment:
todo: nice work getting this tricky nested thing to work. We can't ignore the results of the `random_auxinfo_proof` function, though -- that's the result of our test. You should return the result of that function call instead of `Ok(())`